### PR TITLE
fix(set_upstream): fill node info when upstream node update

### PR DESCRIPTION
**Pull request by ZhangShangyu**
### Description

Fixes https://github.com/apache/apisix/issues/9805

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by shreemaan-abhishek** [comment]
Could you please add test cases?

**Comment by ZhangShangyu** [comment]
i have no good idea to write test about this scene, any suggestions?
btw i am stuck at setup and run test framework on my m1 macbook😂

**Comment by leslie-tsang** [comment]
> i have no good idea to write test about this scene, any suggestions? btw i am stuck at setup and run test framework on my m1 macbook😂

@shreemaan-abhishek cloud please help him to add test case to this PR ? :)

**Comment by shreemaan-abhishek** [comment]
@leslie-tsang I haven't repro the bug yet. Let me try to do that first.

**Comment by Sn0rt** [comment]
we need test case to ensure the https://github.com/apache/apisix/issues/9805 has been fixed.

**Comment by shreemaan-abhishek** [comment]
@leslie-tsang, @ZhangShangyu's point is valid and the fix looks correct to me. But I don't know how we can provide a simpler way to test this case.

The current way to implement testing would be to create a discovery plugin as proposed by the user (the plugin is for an extremely specific purpose) and then making some requests.

**Comment by shreemaan-abhishek** [comment]
Github actions shows workflows are in progress even if they have completed running:
image

I will close and re-open this PR to try to fix this issue.

**Comment by ZhangShangyu** [comment]
> @leslie-tsang, @ZhangShangyu's point is valid and the fix looks correct to me. But I don't know how we can provide a simpler way to test this case.
> 
> The current way to implement testing would be to create a discovery plugin as proposed by the user (the plugin is for an extremely specific purpose) and then making some requests.

the root cause is the cache inconsistent, how about write test to verify the cache?

**Comment by shreemaan-abhishek** [comment]
@ZhangShangyu, looks like your changes broke something. Please fix the CI failure:

image
https://github.com/apache/apisix/actions/runs/5851484873/job/15862317840?pr=9806#step:19:515

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu, looks like your changes broke something. Please fix the CI failure:
> 
> image https://github.com/apache/apisix/actions/runs/5851484873/job/15862317840?pr=9806#step:19:515

it like the config validation fail, my change seems not related to the validation, can rerun this test?
image

**Comment by shreemaan-abhishek** [comment]
@ZhangShangyu, I am convinced that your changes are producing the CI failures. The re-run also failed.

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu, I am convinced that your changes are producing the CI failures. The re-run also failed.

i saw this PR check job also failed, but it was fixed now, can i know how do you fix it?
https://github.com/apache/apisix/actions/runs/5857369728/job/15879091957?pr=10008

**Comment by shreemaan-abhishek** [comment]
> i saw this PR check job also failed, but it was fixed now, can i know how do you fix it?

We re-run the CI. I did it for your PR as well but it didn't work.

**Comment by Sn0rt** [comment]
@ZhangShangyu can you provider the permisson for cooperation with @shreemaan-abhishek

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu can you provider the permisson for cooperation with @shreemaan-abhishek

you mean add collaborators in my repo? if so i added just now

**Comment by shreemaan-abhishek** [comment]
@ZhangShangyu you are using the service discovery feature essentially for routing, your use case is not a standard practice and the bugfix is breaking other things. Are you able to reproduce this bug using existing service discoverers?

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu you are using the service discovery feature essentially for routing, your use case is not a standard practice and the bugfix is breaking other things. Are you able to reproduce this bug using existing service discoverers?

hi, the point is that upstream nodes change frequently and change back like A -> B -> A,  
using existing service discoverer like Nacos or others if has this discover case it will also cause the unexpected behavior

anyway let me try to run and fix the failure tests, thx😄

**Comment by shreemaan-abhishek** [comment]
> hi, the point is that upstream nodes change frequently and change back like A -> B -> A,

in the real world, it is very unlikely that the upstream nodes would change frequently like that.

**Comment by monkeyDluffy6017** [comment]
@ZhangShangyu are you still handling this issue?

**Comment by monkeyDluffy6017** [comment]
@Sn0rt It's a cache problem, please help to handle this

**Comment by Revolyssup** [comment]
I have tested this PR. And it LGTM. But we will have to add tests for it using existing service discovery mechanisms.

**Comment by shreemaan-abhishek** [comment]
Github action isn't allowing re-runs, closing and reopening this PR because it might fix the issue.

**Comment by Revolyssup** [comment]
@ZhangShangyu Can you rebase with master?

File: apisix/upstream.lua [hunk 1]
```diff
@@ -282,6 +282,19 @@ function _M.set_by_route(route, api_ctx)
             end
 
             up_conf.nodes = new_nodes
+            -- fill node info before clone
+            if not is_http then
+                local ok, err = fill_node_info(up_conf, nil, true)
+                if not ok then
+                    return 503, err
```
Comment (Revolyssup) [inline apisix/upstream.lua:289]: can you change this to 502? This is what's failing the test

File: apisix/upstream.lua [hunk 2]
```diff
@@ -282,6 +282,19 @@ function _M.set_by_route(route, api_ctx)
...
+                end
+            else
+                local ok, err = fill_node_info(up_conf, api_ctx.upstream_scheme, false)
+                if not ok then
+                    return 503, err
```
Comment (Revolyssup) [inline apisix/upstream.lua:294]: dittto

**Review by Revolyssup — CHANGES_REQUESTED** [review]

**Comment by Revolyssup** [comment]
@ZhangShangyu If you don't have time, you can give me permissions on this branch and I can take this over.

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu If you don't have time, you can give me permissions on this branch and I can take this over.

ok, just in holiday, thank you

**Comment by ZhangShangyu** [comment]
> @ZhangShangyu Can you rebase with master?

done

**Comment by monkeyDluffy6017** [comment]
@Revolyssup @ZhangShangyu The ci is still failing, do you guys want to keep working on this?

**Comment by Revolyssup** [comment]
> @Revolyssup @ZhangShangyu The ci is still failing, do you guys want to keep working on this?

I'll take a look here and update today.

**Comment by ZhangShangyu** [comment]
> > @Revolyssup @ZhangShangyu The ci is still failing, do you guys want to keep working on this?
> 
> I'll take a look here and update today.

Hi @Revolyssup, any update?

**Comment by Revolyssup** [comment]
> > > @Revolyssup @ZhangShangyu The ci is still failing, do you guys want to keep working on this?
> > 
> > 
> > I'll take a look here and update today.
> 
> Hi @Revolyssup, any update?

Yeah I am looking at the changes right now. I had gotten busy with other stuff previously. Failing tests are related to changes. I will ping in case I need help

**Comment by sheharyaar** [comment]
@Revolyssup , I would like to take up this PR.

**Comment by monkeyDluffy6017** [comment]
It's fixed by https://github.com/apache/apisix/pull/10722

**Pull request closed without merging**